### PR TITLE
logic for converting Circle workflow events into a chart #7

### PR DESCRIPTION
**Pull request by claymccoy**
logic for converting Circle workflow events into our flat stage abstraction and render it into a chart
this needs to be integrated into an event handler somewhere that listens for the workflow events and then the chart url put in a slack attachment

### CODE REVIEW

**Comment by claymccoy** [comment]
Here is a real execution of this logic to render a chart of the workflow that we keep using as an example:
https://circleci.com/workflow-run/55bac556-40c0-4f4f-86b2-2b35b7064112
screen shot 2017-11-06 at 10 35 05 am

**Comment by ddgenome** [comment]
Looks like you need to delint: `npm run lint:fix`

File: src/handlers/event/build/workflow/ChartUrl.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    let scale = 1000;
+    if (longestDuration > 600000) {
+        scaleText = "in minutes";
+        scale = 60000;
+    }
+    const chartHeight = 100 + stages.length * 10;
+
+    const durationPlaceholder = _.round((longestDuration * .02) / scale);
+
+    const longestDurationsInMinutes = [];
```
Comment (ddgenome) [inline src/handlers/event/build/workflow/ChartUrl.ts]: Might be useful to include types in these declarations.

File: src/handlers/event/build/workflow/CircleWorkflow.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+                }
+                stages.push(stage);
+            } else {
+                stage.jobs.push(name);
+            }
+        }
+    })
+
+    const orderedStages: Stage[] = [{
+        jobs: ["build"],
```
Comment (ddgenome) [inline src/handlers/event/build/workflow/CircleWorkflow.ts]: What is special about `"build"`?

File: src/handlers/event/build/workflow/CircleWorkflow.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    const workflowStages: WorkflowStage[] = _.map(orderedStages,s => {
+        const buildsInStage = workflow.builds.filter(b => _.includes(s.jobs, b.jobName))
+        let jobName = s.jobs[0];
+        if (buildsInStage.length == 0) {
+            return { name: jobName } as WorkflowStage
+        }
+        let longestJobDuration = 0;
+        let earliestStart = undefined;
```
Comment (ddgenome) [inline src/handlers/event/build/workflow/CircleWorkflow.ts]: Setting to `undefined` is unnecessary. Better to just specify a type.

**Comment by claymccoy** [comment]
I delinted manually. It sucked. We should just run "npm run lint:fix" autmatically and add it to the PR. That is kind of the Atomist thing to do.

File: src/handlers/event/build/workflow/ChartUrl.ts (hunk 1 above)
Comment (claymccoy) [reply src/handlers/event/build/workflow/ChartUrl.ts]: okay

**Comment by ddgenome** [comment]
@jessitron has an automation that does that. I just have it build in to my editor.

**Comment by jessitron** [comment]
yeah I just need to get cd's lint automation in docker-samples running in prod. maybe I'll do that right this second because I keep running into it too

**Pull request merged**